hypothesis-python/tests/conjecture/test_ir_tree.py

```diff
@@ -0,0 +1,224 @@
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Copyright the Hypothesis Authors.
+# Individual contributors are listed in AUTHORS.rst and the git log.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+import pytest
+
+from hypothesis import given, settings, strategies as st
+from hypothesis.internal.conjecture.data import IRTree, IRTreeLeaf
+from hypothesis.internal.floats import SMALLEST_SUBNORMAL, float_to_int
+from hypothesis.internal.intervalsets import IntervalSet
+from hypothesis.stateful import RuleBasedStateMachine, invariant, rule
+
+from tests.conjecture.common import (
+    draw_boolean_kwargs,
+    draw_bytes_kwargs,
+    draw_float_kwargs,
+    draw_integer_kwargs,
+    draw_string_kwargs,
+    fresh_data,
+    ir_types_and_kwargs,
+)
+
+
+def draw_ir_value(ir_type, kwargs):
+    data = fresh_data()
+    return getattr(data, f"draw_{ir_type}")(**kwargs)
+
+
+def ir_value_eq(ir_type, v1, v2):
+    if ir_type == "float":
+        # correctly handle -0.0 == 0.0 and nan == nan
+        return float_to_int(v1) == float_to_int(v2)
+    return v1 == v2
+
+
+@st.composite
+def leaves(draw, *, was_forced=None):
+    (ir_type, kwargs) = draw(ir_types_and_kwargs())
+    return IRTreeLeaf(
+        ir_type=ir_type,
+        value=draw_ir_value(ir_type, kwargs),
+        kwargs=kwargs,
+        was_forced=was_forced if was_forced is not None else draw(st.booleans()),
+    )
+
+
+@given(leaves())
+def test_copy_leaf_is_equal(leaf):
+    assert leaf == leaf.copy()
+
+
+@given(leaves(was_forced=False))
+def test_copy_leaf_with_value(leaf):
+    ir_type = leaf.ir_type
+    new_value = draw_ir_value(ir_type, leaf.kwargs)
```

Comment:
Nondeterminism alert!  When using `fresh_data()` inside an existing property-based test, you'll need to pass in a hypothesis-controlled (or at least deterministically-seeded) `Random` instance (e.g. from `st.randoms()`); otherwise shrinking etc. will be flaky.

I'd suggest adding an assertion to `fresh_data()` or `draw_ir_value()` to check that `random=` is passed if you're in a `build_context`.  Alternatively I think you could grab the current `ConjectureData` off the context, draw a `Random` instance, and add it - I think that level of magic is OK in tests and it's certainly ergonomic...

Reply:
I've definitely been abusing `fresh_data()` here 😅.